data/transactions/logic/assembler_test.go

```diff
@@ -2421,3 +2438,72 @@ func TestBadInnerFields(t *testing.T) {
 	testProg(t, "itxn_begin; byte 0x7263; itxn_field VotePK", 6)
 	testProg(t, "itxn_begin; int 32; bzero; itxn_field TxID", 6, Expect{4, "...is not allowed."})
 }
+
+func TestTypeTracking(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	testProg(t, "+", LogicVersion, Expect{1, "+ expects 2 stack arguments..."})
+	// hitting a label in deadcode starts analyzing again, with unknown stack
+	testProg(t, "b end; label: +; end: b label", LogicVersion)
+	// callsub also wipes our stack knowledge, this tests shows why: it's properly typed
+	testProg(t, "callsub A; +; return; A: int 1; int 2; retsub", LogicVersion)
+	// retsub deadens code, like any unconditional branch
+	testProg(t, "callsub A; +; return; A: int 1; int 2; retsub; concat", LogicVersion)
+
+	// Branching would have confused the old analysis, but the problem is local
+	// to a basic block, so it makes sense to report it.
+	testProg(t, `
+ int 1
+ b confusion
+label:
+ byte "john"					// detectable mistake
+ int 2
+ +
+confusion:
+ b label
+`, LogicVersion, Expect{7, "+ arg 0 wanted type uint64..."})
+
+	// Unless that same error is in dead code.
+	testProg(t, `
+ int 1
+ b confusion
+label:
+ err							// deadens the apparent error at +
+ byte "john"
+ int 2
+ +
+confusion:
+ b label
+`, LogicVersion)
+
+	// Unconditional branches also deaden
+	testProg(t, `
+ int 1
+ b confusion
+label:
+ b done							// deadens the apparent error at +
+ byte "john"
+ int 2
+ +
+confusion:
+ b label
+done:
+`, LogicVersion)
+
+	// Turning type tracking off and then back on, allows any follow-on code.
+	testProg(t, `
+ int 1
+ int 2
+#pragma typetrack false
+ concat
+`, LogicVersion)
+
+	testProg(t, `
```

Comment:
@jannotti I think there's an additional test case worth considering.  I opened https://github.com/jannotti/go-algorand/pull/6 to help illustrate.  Can you have a look?

Reply:
@jannotti Thanks for taking in https://github.com/jannotti/go-algorand/pull/6.  Resolving.